### ISSUE TITLE
[Improvement]: Make the type engine to work when part of the type is SemType and the other part is BType

### ISSUE THREAD

**Issue by lochana-chathura** [ballerina-platform/ballerina-lang#41066]
### Description

$subject.

### Describe your problem(s)

In the current integration of sem-types, we can't ship it until we fully complete the sem-type engine for all the types.
Therefore, as per our recent discussions we need to make the type checker work when only part of the type are sem-types.  

### Describe your solution(s)
A type would have two components. A `BType` component and a `SemType` component.
When we do type operations like union, intersection, and subtyping, we can split it into two sub-problems. 

Let type P = P1|P2 and Q = Q1|Q2.
Here P1/Q1 are the BType components and P2/Q2 are the semtype components. We expect BType and SemType components to be disjoint. 

Then we can derive the operations on top of P and Q like this. 

- Subtyping: P <: Q => (P1 <: Q1) & (P2 <: Q2)
- Intersection: P & Q => (P1 & Q1) | (P2 & Q2)
- Union: P | Q => (P1 | Q1) | (P2 | Q2)
- Complement: !P => !P1 & !P2

In the above, each operation is split into two operations where the first operation is done from the existing logic whereas the second operation is done from the new logic.

**Comment by lochana-chathura** [comment]
Update:

- Converted all the **Nil, Boolean, Int, String, Decimal, and Float** type value spaces into SemTypes and used them in the `isAssigable()`. Other usages still have to be converted. 

- Current status of JBallerin unit tests: _**8415 tests completed, 67 failed, 89 skipped**_

**Comment by lochana-chathura** [comment]
Update:

- Fixed `isAssigable()` to handle recursive cases as well with semtypes. Failed tests were down to 9. They are due to edge cases when resolving `BFiniteType`.
- Converting other usages of BFiniteType is in progress. `BFiniteType` was removed from `JvmTypeGen`.
- Current status of JBallerin unit tests: **8414 tests completed, 16 failed, 44 skipped**

**Comment by lochana-chathura** [comment]
Update:
- Converted all the usages of BFiniteType `valueSpace` (which was modeled as a set of expressions) to use resolved sem-type. There were about 90 usages.
- Fixed the toString() representation for user diagnostics as the sem-type-based toString() representation is in a normalized form. We are preserving the original user written toString() at the type resolving. 
- Current status of JBallerin unit tests: **8418 tests completed, 31 failed, 237 skipped**

**Comment by lochana-chathura** [comment]
This issue had been parked for a while due to eggplant priorities. However, it is now back in progress.

Current state:
All tests are passing except for the tests that rely upon the syntax order of the union type. e.g. cloneWithType
After a discussion with James, we need to maintain the order of the basic type bitset at least.

**Comment by lochana-chathura** [comment]
> Current state: All tests are passing except for the tests that rely upon the syntax order of the union type. e.g. cloneWithType After a discussion with James, It was concluded to maintain the order of the basic type bitset at least.

Tests are all passing with the fix done with #42158.

Current state: Working on refactoring several places before sending the final PR.

**Comment by lochana-chathura** [comment]
Changes are now complete.

Following performance tests were conducted on top of the changes.

1. Building the nBallerina compiler
2. Running jBallerina unit tests

Results are attached with the screenshot.

<img width="1555" alt="Screenshot 2024-03-05 at 16 54 38" src="https://github.com/ballerina-platform/ballerina-lang/assets/39232462/f082a061-dbeb-4923-be85-c0ecff57b60a">

In summary, there is a 2% performance drop with the changes. This could be because we are populating an extra sem-type property for each BType.

**Comment by lochana-chathura** [comment]
As discussed with @hasithaa, it was decided to not include these changes for update 9. However, will be shipped in a future release. 

Closing the issue as changes are done and are available at the [nutcracker](https://github.com/ballerina-platform/ballerina-lang/tree/nutcracker) branch.
